src/monster.cpp

```diff
@@ -126,18 +126,18 @@ void Monster::onCreatureAppear(Creature* creature, bool isLogin)
 		lua_State* L = scriptInterface->getLuaState();
 		scriptInterface->pushFunction(mType->info.creatureAppearEvent);
 
-		tfs::lua::pushUserdata(L, this);
+		tfs::lua::pushSharedPtr(L, getMonster());
 		tfs::lua::setMetatable(L, -1, "Monster");
 
-		tfs::lua::pushUserdata(L, creature);
+		tfs::lua::pushSharedPtr(L, creature);
 		tfs::lua::setCreatureMetatable(L, -1, creature);
 
 		if (scriptInterface->callFunction(2)) {
 			return;
 		}
 	}
 
-	if (creature == this) {
+	if (creature.get() == this) {
```

Comment:
Could this be `const auto`
```suggestion
	if (creature == getMonster()) {
```